src/sentry/eventstore/snuba/backend.py

```diff
@@ -108,9 +110,62 @@ def __get_events(
         should_bind_nodes=False,
     ):
         assert filter, "You must provide a filter"
-        cols = self.__get_columns(additional_columns)
+        cols = self.__get_columns()
         orderby = orderby or DESC_ORDERING
 
+        # This is an optimization for the Group.filter_by_event_id query where we
+        # have a single event ID and want to check all accessible projects for a
+        # direct hit. In this case it's usually faster to go to nodestore first.
+        if (
+            filter.event_ids
+            and filter.project_ids
+            and len(filter.event_ids) * len(filter.project_ids) < min(limit, NODESTORE_LIMIT)
+            and offset == 0
+        ):
+            event_list = [
+                Event(project_id=project_id, event_id=event_id)
+                for event_id in filter.event_ids
+                for project_id in filter.project_ids
+            ]
+            self.bind_nodes(event_list)
+
+            nodestore_events = [event for event in event_list if len(event.data)]
+
+            if nodestore_events:
+                event_ids = {event.event_id for event in nodestore_events}
+                project_ids = {event.project_id for event in nodestore_events}
+                start = min(event.datetime for event in nodestore_events)
+                end = max(event.datetime for event in nodestore_events) + timedelta(seconds=1)
+
+                result = snuba.aliased_query(
+                    selected_columns=cols,
+                    start=start,
+                    end=end,
+                    conditions=filter.conditions,
+                    filter_keys={"project_id": project_ids, "event_id": event_ids},
+                    orderby=orderby,
+                    limit=len(nodestore_events),
+                    offset=DEFAULT_OFFSET,
+                    referrer=referrer,
+                    dataset=snuba.Dataset.Events,
+                )
+
+                if "error" not in result:
+                    events = [self.__make_event(evt) for evt in result["data"]]
+
+                    # Bind previously fetched node data
+                    for event in events:
+                        node_data = next(
+                            e
+                            for e in nodestore_events
+                            if event.event_id == e.event_id and event.project_id == e.project_id
+                        ).data.data
```

Comment:
This way it would be easier to understand what this code is trying to do.

> filter(lambda n: n.event_id == event.event_id and n.project_id == event.project_id, nodestore_events)

But better (more clear and avoiding one lsit scan per event):
```
nodestore_dict = {(e.event_id, e.project_id): e for e in nodestore_events}
events = [
    e.data.bind_data(nodestore_dict[(e.event_id, e.project_id)].data.data) for e in events
]
```

Reply:
☝️ I'll update to use a dict, I'm going to assume you don't mean to actually reassign "events" there

Reply:
ok, yes, without the reassignment